server/src/main/java/io/deephaven/server/arrow/BarrageSubscriptionRequestHandler.java

```diff
@@ -0,0 +1,216 @@
+//
+// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.server.arrow;
+
+import com.google.auto.service.AutoService;
+import com.google.rpc.Code;
+import io.deephaven.barrage.flatbuf.BarrageMessageType;
+import io.deephaven.barrage.flatbuf.BarrageSubscriptionRequest;
+import io.deephaven.engine.liveness.LivenessReferent;
+import io.deephaven.engine.table.impl.perf.QueryPerformanceNugget;
+import io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder;
+import io.deephaven.extensions.barrage.BarrageMessageWriter;
+import io.deephaven.extensions.barrage.BarrageSubscriptionOptions;
+import io.deephaven.extensions.barrage.util.BarrageProtoUtil;
+import io.deephaven.extensions.barrage.util.GrpcUtil;
+import io.deephaven.proto.util.Exceptions;
+import io.deephaven.proto.util.ExportTicketHelper;
+import io.deephaven.server.session.SessionState;
+import io.deephaven.server.session.TicketRouter;
+import io.deephaven.util.SafeCloseable;
+import io.grpc.stub.StreamObserver;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Handler for BarrageSubscriptionRequest over DoExchange.
+ */
+public class BarrageSubscriptionRequestHandler implements ArrowFlightUtil.DoExchangeMarshaller.Handler {
+
+    private final ArrowFlightUtil.DoExchangeMarshaller marshaller;
+    private final TicketRouter ticketRouter;
+    private final SessionState session;
+    private final StreamObserver<BarrageMessageWriter.MessageView> listener;
+
+    /**
+     * The object representing the subscription.
+     */
+    private ExchangeMarshaller.Subscription subscriptionObject;
+
+    private Queue<BarrageSubscriptionRequest> preExportSubscriptions;
+    private SessionState.ExportObject<?> onExportResolvedContinuation;
+
+    public BarrageSubscriptionRequestHandler(final ArrowFlightUtil.DoExchangeMarshaller marshaller,
+            final TicketRouter ticketRouter,
+            final SessionState session,
+            final StreamObserver<BarrageMessageWriter.MessageView> listener) {
+        this.marshaller = marshaller;
+        this.ticketRouter = ticketRouter;
+        this.session = session;
+        this.listener = listener;
+    }
+
+    @Override
+    public void handleMessage(@NotNull final BarrageProtoUtil.MessageInfo message) {
+        validateMessage(message);
+
+        final BarrageSubscriptionRequest subscriptionRequest = BarrageSubscriptionRequest
+                .getRootAsBarrageSubscriptionRequest(message.app_metadata.msgPayloadAsByteBuffer());
+
+        if (subscriptionObject != null) {
+            apply(subscriptionRequest);
+            return;
```

Comment:
I think we also want to guard this check, so that we don't skate past the chance to use the existing subscription, fail to enqueue, and re-create the subscription?